File: zak/core/runtime/agent.py
"""
ZAK Runtime — BaseAgent, AgentContext, AgentResult.

All security agents in ZAK inherit from BaseAgent and receive an AgentContext
at execution time. Results are always wrapped in AgentResult.
"""

from __future__ import annotations

import abc
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from zak.core.dsl.schema import AgentDSL


@dataclass
class AgentContext:
    """
    Runtime context injected into every agent execution.

    Carries tenant identity, trace ID, and the validated DSL for the agent.
    This is the single source of truth for 'who is running, under which tenant,
    with what permissions'.
    """
    tenant_id: str
    trace_id: str
    dsl: AgentDSL
    environment: str = "staging"
    metadata: dict[str, Any] = field(default_factory=dict)
    started_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    @property
    def agent_id(self) -> str:
        return self.dsl.agent.id


@dataclass
class AgentResult:
    """Typed result envelope returned by every agent execution."""
    success: bool
    agent_id: str
    tenant_id: str
    trace_id: str
    output: dict[str, Any] = field(default_factory=dict)
    errors: list[str] = field(default_factory=list)
    duration_ms: float = 0.0
    completed_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    @classmethod
    def ok(
        cls,
        context: AgentContext,
        output: dict[str, Any],
        duration_ms: float = 0.0,
    ) -> AgentResult:
        return cls(
            success=True,
            agent_id=context.agent_id,
            tenant_id=context.tenant_id,
            trace_id=context.trace_id,
            output=output,
            duration_ms=duration_ms,
        )

    @classmethod
    def fail(
        cls,
        context: AgentContext,
        errors: list[str],
        duration_ms: float = 0.0,
    ) -> AgentResult:
        return cls(
            success=False,
            agent_id=context.agent_id,
            tenant_id=context.tenant_id,
            trace_id=context.trace_id,
            errors=errors,
            duration_ms=duration_ms,
        )


class BaseAgent(abc.ABC):
    """
    Abstract base class for all ZAK security agents.

    Every agent must implement:
    - execute(context) → AgentResult

    Optionally override:
    - pre_run(context)  — setup / validation before main execution
    - post_run(context, result) — cleanup / result enrichment
    """

    @property
    def name(self) -> str:
        return self.__class__.__name__

    def pre_run(self, context: AgentContext) -> None:
        """Optional hook called before execute(). Override for setup logic."""
        pass

    @abc.abstractmethod
    def execute(self, context: AgentContext) -> AgentResult:
        """
        Core agent logic. Must be implemented by every concrete agent.

        Args:
            context: Runtime context with tenant, trace ID, and DSL.

        Returns:
            AgentResult with success status and typed output.
        """

    def post_run(self, context: AgentContext, result: AgentResult) -> None:
        """Optional hook called after execute(). Override for cleanup logic."""
        pass

File: adk/node/src/core/runtime/agent.ts
/**
 * ZAK Runtime — BaseAgent, AgentContext, AgentResult.
 *
 * All security agents in ZAK inherit from BaseAgent and receive an AgentContext
 * at execution time. Results are always wrapped in AgentResult.
 *
 * TypeScript equivalent of zak/core/runtime/agent.py.
 */

import type { AgentDSL } from "../dsl/schema.js";

// ---------------------------------------------------------------------------
// AgentContext
// ---------------------------------------------------------------------------

/**
 * Runtime context injected into every agent execution.
 *
 * Carries tenant identity, trace ID, and the validated DSL for the agent.
 * This is the single source of truth for 'who is running, under which tenant,
 * with what permissions'.
 */
export interface AgentContext {
  tenantId: string;
  traceId: string;
  dsl: AgentDSL;
  environment: string;
  metadata: Record<string, unknown>;
  startedAt: Date;
}

/**
 * Create an AgentContext with defaults.
 */
export function createAgentContext(opts: {
  tenantId: string;
  traceId: string;
  dsl: AgentDSL;
  environment?: string;
  metadata?: Record<string, unknown>;
}): AgentContext {
  return {
    tenantId: opts.tenantId,
    traceId: opts.traceId,
    dsl: opts.dsl,
    environment: opts.environment ?? "staging",
    metadata: opts.metadata ?? {},
    startedAt: new Date(),
  };
}

/**
 * Get the agent ID from a context (convenience accessor).
 */
export function getAgentId(ctx: AgentContext): string {
  return ctx.dsl.agent.id;
}

// ---------------------------------------------------------------------------
// AgentResult
// ---------------------------------------------------------------------------

/**
 * Typed result envelope returned by every agent execution.
 */
export interface AgentResult {
  success: boolean;
  agentId: string;
  tenantId: string;
  traceId: string;
  output: Record<string, unknown>;
  errors: string[];
  durationMs: number;
  completedAt: Date;
}

/**
 * Create a successful AgentResult.
 */
export function agentResultOk(
  context: AgentContext,
  output: Record<string, unknown>,
  durationMs = 0
): AgentResult {
  return {
    success: true,
    agentId: getAgentId(context),
    tenantId: context.tenantId,
    traceId: context.traceId,
    output,
    errors: [],
    durationMs,
    completedAt: new Date(),
  };
}

/**
 * Create a failed AgentResult.
 */
export function agentResultFail(
  context: AgentContext,
  errors: string[],
  durationMs = 0
): AgentResult {
  return {
    success: false,
    agentId: getAgentId(context),
    tenantId: context.tenantId,
    traceId: context.traceId,
    output: {},
    errors,
    durationMs,
    completedAt: new Date(),
  };
}

// ---------------------------------------------------------------------------
// BaseAgent
// ---------------------------------------------------------------------------

/**
 * Abstract base class for all ZAK security agents.
 *
 * Every agent must implement:
 * - execute(context) → AgentResult
 *
 * Optionally override:
 * - preRun(context)  — setup / validation before main execution
 * - postRun(context, result) — cleanup / result enrichment
 */
export abstract class BaseAgent {
  /**
   * Human-readable agent name (defaults to constructor name).
   */
  get name(): string {
    return this.constructor.name;
  }

  /**
   * Optional hook called before execute(). Override for setup logic.
   */
  preRun(_context: AgentContext): void {
    // No-op by default
  }

  /**
   * Core agent logic. Must be implemented by every concrete agent.
   */
  abstract execute(context: AgentContext): AgentResult | Promise<AgentResult>;

  /**
   * Optional hook called after execute(). Override for cleanup logic.
   */
  postRun(_context: AgentContext, _result: AgentResult): void {
    // No-op by default
  }
}
